usr/src/cmd/pkgsrc-wrapper/pkgsrc-wrapper.c

```diff
@@ -0,0 +1,181 @@
+static const char *pkgsrc_paths[] = {
+	"/opt/local/bin",
+	"/opt/local/sbin",
+	"/opt/tools/bin",
+	"/opt/tools/sbin",
+	NULL
+};
+
+/*
+ * This returns an allocated string.  Since we're exec()-ing or exit()-ing we
+ * really don't need to worry about leaks, however.
+ */
+static char *
+generate_pkgsrc_path(const char *desired_bin)
+{
+	const char *path = NULL;
+	char *binpath;
+	uint_t index;
+
+	for (index = 0; pkgsrc_paths[index] != NULL; index++) {
+		int dirfd;
+		int cmdfd;
+
+		/*
+		 * NOTE: Any of dirfd or cmdfd that has an actual file
+		 * descriptor from open(2) MUST be closed before this loop
+		 * either exits or iterates.
+		 */
+
+		dirfd = open(pkgsrc_paths[index], O_RDONLY);
+		if (dirfd == -1) {
+			/* XXX KEBE ASKS, print diagnostics? */
+			continue; /* Try next one */
+		}
+		cmdfd = openat(dirfd, desired_bin, O_RDONLY);
+		(void) close(dirfd); /* XXX KEBE WARNS, errno reset... */
+		if (cmdfd == -1) {
+			/*
+			 * XXX KEBE ASKS, print diagnostics?
+			 * If so, move close(dirfd) here after printing, or
+			 * save off errno for openat() above.
+			 */
+			continue; /* Try next one */
+		}
+		(void) close(cmdfd);
+		path = pkgsrc_paths[index];
+		break;	/* We're done. */
+	}
+
+	if (path == NULL)
+		return (NULL); /* If we reach here, there's no such binary. */
+
+	/*
+	 * Okay, so we can now combine path and desired_bin into a string
+	 * suitable for an exec()-family call.
+	 */
+
+	if (asprintf(&binpath, "%s/%s", path, desired_bin) == -1) {
+		/* XXX KEBE ASKS, print diagnostics? */
+		assert(binpath == NULL); /* man pages says this is true. */
+	}
+
+	return (binpath);
+}
+
+/* Don't let us run ourselves. */
+static char *myname = "pkgsrc-wrapper";
+
+int
+main(int argc, char *argv[], char *envp[])
+{
+	const char *base_bin;
+	const char *remapped_bin;
+	char *pkgsrc_path;
+
+	/*
+	 * 1. Get the binary name in argv[0], remapping if need be.
+	 * (e.g. csh -> tcsh, or -($SHELL) invocations.)
+	 */
+	base_bin = strrchr(argv[0], '/');
+	if (base_bin == NULL)
+		base_bin = argv[0];
+	else
+		base_bin++; /* Move past matched '/' */
+
+	/*
+	 * 2. Make sure we just aren't running ourselves!
+	 * (Can use strcmp() because "myname" is bounded.)
+	 */
+	if (strcmp(myname, base_bin) == 0) {
+		/* XXX KEBE ASKS L10N? */
+		(void) fprintf(stderr, "%s should not be run by itself.\n",
+		    myname);
+		exit(3);
+	}
+
+	remapped_bin = remap_bin(base_bin);
+
+	/* 3. See if the binary name is in one of the pkgsrc paths. */
+	pkgsrc_path = generate_pkgsrc_path(remapped_bin);
+	if (pkgsrc_path == NULL) {
+		/* XXX KEBE ASKS L10N? */
+		/* XXX KEBE ALSO ASKS -- better message? */
+		(void) fprintf(stderr, "Please install %s from pkgsrc.\n",
+		    remapped_bin);
+		exit(1);
+	}
+
+	/* 4. If so, launch it with our exact arg[vc] (assume env goes thru?) */
+	if (execve(pkgsrc_path, argv, envp) == -1) {
+		/* XXX KEBE ASKS L10N? */
+		/* XXX KEBE ALSO ASKS -- better message? */
+		(void) fprintf(stderr, "Cannot execute %s: %s.\n",
```

Comment:
"Failed to execute %s: %s.\n"

"Cannot" seems to imply "I was going to, but something stopped me from trying", whereas failed implies "I tried, but something went wrong"

Reply:
Will fix.